src/main/java/uk/gov/hmcts/dts/fact/services/admin/AdminCourtTypesAndCodesService.java

```diff
@@ -0,0 +1,144 @@
+
+        return new CourtTypesAndCodes(getCourtCourtTypes(courtEntity), courtEntity.getGbs(), getCourtDxCodes(courtEntity));
+    }
+
+    private List<CourtType> getCourtCourtTypes(final Court courtEntity) {
+
+        final List<CourtType> returnCourtTypes = courtEntity.getCourtTypes()
+                .stream()
+                .map(CourtType::new)
+                .collect(toList());
+        return mapCourtCode.mapCourtCodesForCourtTypeModel(returnCourtTypes, courtEntity);
+    }
+
+    public List<DxCode> getCourtDxCodes(final Court courtEntity) {
+        return  courtDxCodesRepository.findByCourtId(courtEntity.getId())
+            .stream()
+            .map(CourtDxCode::getDxCode)
+            .map(DxCode::new)
+            .collect(toList());
+    }
+
+    @Transactional()
+    public CourtTypesAndCodes updateCourtTypesAndCodes(final String slug, final CourtTypesAndCodes courtTypesAndCodes) {
+        final Court courtEntity = courtRepository.findBySlug(slug)
+            .orElseThrow(() -> new NotFoundException(slug));
+        return saveNewCourtTypesAndCodes(courtEntity, courtTypesAndCodes);
+
+    }
+
+    protected CourtTypesAndCodes saveNewCourtTypesAndCodes(final Court courtEntity, final CourtTypesAndCodes courtTypesAndCode) {
+
+        final List<uk.gov.hmcts.dts.fact.entity.CourtType> courtTypeEntities = getNewCourtCourtTypesEntity(courtTypesAndCode.getCourtTypes());
+
+        //save CourtTypesAndGbsCodes
+        final Court amendedCourtEntity = saveCourtTypesAndGbsCodes(courtEntity,courtTypesAndCode,courtTypeEntities);
+
+        //save court dx codes if request is by SuperAdmin
+        if (rolesProvider.getRoles().contains(FACT_SUPER_ADMIN)) {
+            saveCourtDxCodes(amendedCourtEntity, getNewDxCodeEntity(courtTypesAndCode.getDxCodes()));
+        }
+
+        return new CourtTypesAndCodes(getCourtCourtTypes(amendedCourtEntity), amendedCourtEntity.getGbs(), getCourtDxCodes(amendedCourtEntity));
+    }
+
+
+    protected Court saveCourtTypesAndGbsCodes(final Court courtEntity, final CourtTypesAndCodes courtTypesAndCode, final List<uk.gov.hmcts.dts.fact.entity.CourtType> courtTypeEntity) {
+        if (courtEntity.getCourtTypes() == null) {
+            courtEntity.setCourtTypes(courtTypeEntity);
+        } else {
+            courtEntity.getCourtTypes().clear();
+            courtEntity.getCourtTypes().addAll(courtTypeEntity);
+        }
+
+        //remove existing code entries for court.
+        courtEntity.setMagistrateCode(null);
+        courtEntity.setCciCode(null);
+        courtEntity.setNumber(null);
+
+        //set court gbs code only if request is by SuperAdmin
+        if (rolesProvider.getRoles().contains(FACT_SUPER_ADMIN)) {
+            courtEntity.setGbs(courtTypesAndCode.getGbsCode());
+        }
+
+        final Court amendedCourtEntity = mapCourtCode.mapCourtCodesForCourtEntity(courtTypesAndCode.getCourtTypes(), courtEntity);
+
+        return courtRepository.save(amendedCourtEntity);
```

Comment:
Think `save` by default is transactional, so not sure if we need the @transactional annotation at the top, one to bear in mind? Not sure if it'll matter/have any impact on it though

Reply:
remove @transactional annotation